File: Task/Factorial-primes/PascalABC.NET/factorial-primes.pas
function IsPrime(n: int64): boolean;
begin
  if (n = 2) or (n = 3) then Result := true
  else if (n <= 1) or ((n mod 2) = 0) or ((n mod 3) = 0) then Result := false
  else
  begin
    var i := 5;
    Result := False;
    while i <= trunc(sqrt(n)) do
    begin
      if ((n mod i) = 0) or ((n mod (i + 2)) = 0) then exit;
      i += 6;
    end;
    Result := True;
  end;
end;

function Factorial(n: integer): int64;
begin
  Result := 1;
  for var i := 2 to n do Result *= i;
end;

begin
  var found := 0;
  var i := 1;
  while found < 10 do
  begin
    var fact := Factorial(i);
    if IsPrime(fact - 1) then
    begin
      writeln(i:2, '! - 1 = ', fact - 1);
      found += 1;
    end;
    if IsPrime(fact + 1) then
    begin
      writeln(i:2, '! + 1 = ', fact + 1);
      found += 1;
    end;
    i += 1;
  end;
end.

File: Task/Factorial-primes/Haskell/factorial-primes.hs
isPrime :: Integer -> Bool
isPrime n
  | n < 2 = False
  | n == 2 = True
  | otherwise = 0 == length [i | i <- [2..floor(sqrt (fromIntegral n))], n `mod` i == 0]

data Op = SubtractOne | AddOne
  deriving (Show)

data Prime = Prime
  { fact  :: Integer,
    prime :: Integer,
    op    :: Op
} deriving (Show)

getFactorialPrimes :: Integer -> [Prime]
getFactorialPrimes n
  | isPrime (f - 2) = Prime {fact = n, prime = f - 2, op = SubtractOne} :
    (if isPrime f then [Prime {fact = n, prime = f, op = AddOne}] else [])
  | otherwise = (if isPrime f then [Prime {fact = n, prime = f, op = AddOne}] else [])
  where
    factorial n = product[1..n]
    f = factorial n + 1

factorialPrimes = concat $ map getFactorialPrimes [1..]
main = mapM_ print $ take 10 factorialPrimes
